### PR TITLE
Files relates to SLES 15 SP 3 test container - podman

### PR DESCRIPTION
**Pull request by rumch-se**
#### Description:

- _SLES 15 SP3 - contaner - podman_

#### Rationale:

- To automatize test process for SLES 15 SP3

### CODE REVIEW

**Comment by rumch-se** [comment]
Please take in mind the following
1/
In directory /content/Dockerfiles we have:
sles15-sp3v2 - a docker file to build a container 
sshd_config - a configuration file for sshd server - inside the container
public keys
id_dsa_test_key.pub 
id_rsa_test_key.pub
id_ecdsa_test_key.pub
private keys
id_dsa_test_key 
id_rsa_test_key
id_ecdsa_test_key

2/To build an image in Dockerfiles I used:
podman build -t test_suite -f sle15-sp3v2 .

3/To see the container console I used
podman run -ti --rmi test_suite bash

4/ I executed the following command from directory /content/tests
./automatus.py rule --container test_suite --datastream ../build/ssg-sle15-ds-1.2.xml accounts_password_all_shadowed_sha512
.. and **the attached file** shows the results
also as root
sudo  ./automatus.py rule --container test_suite --datastream ../build/ssg-sle15-ds-1.2.xml accounts_password_all_shadowed_sha512
[logs_from_tests.txt](https://github.com/ComplianceAsCode/content/files/9852132/logs_from_tests.txt)

5/The log file in /test/logs/ were empty

**Notes:** 
1/I use as a host os - Ubuntu 20.04
2/I attached the keys if you want to simulate my case - and they were created because of tests

**Review by marcusburghardt — COMMENTED** [review]
I wonder if it wouldn't make sense to create a separate folder for these non-Dockerfiles. That way, they don't get mixed up with Dockerfiles. For organizational purposes only.

**Comment by rumch-se** [comment]
Hello @marcusburghardt

Thank you for your comment. Your are right that some of the files should not be in Dockerfiles directory. I have opened the pull request as a only option to find a solution of my issue from you team - i.e. the image was created successfully, ... and after when I tried to use ./automatus.py - it did not work. The files which are "not docker type" are keys and a config file for ssh server. We use keys to establish a ssh connection without password during the tests, and because this part did not work in my case I decided to reconfigure ssh server.  I can put files in a different locations and I can change the code used by podman (the docker file) during the creation of the image, but this will not change the fact - that I will not able to use ./automatus.py as expected.

Have a nice day
Rumen

**Comment by marcusburghardt** [comment]
> Hello @marcusburghardt
> 
> Thank you for your comment. Your are right that some of the files should not be in Dockerfiles directory. I have opened the pull request as a only option to find a solution of my issue from you team - i.e. the image was created successfully, ... and after when I tried to use ./automatus.py - it did not work. The files which are "not docker type" are keys and a config file for ssh server. We use keys to establish a ssh connection without password during the tests, and because this part did not work in my case I decided to reconfigure ssh server. I can put files in a different locations and I can change the code used by podman (the docker file) during the creation of the image, but this will not change the fact - that I will not able to use ./automatus.py as expected.
> 
> Have a nice day Rumen

Hi @rumch-se , I understand this is not related to the issue you reported. I only commented to ensure this is taken in consideration when the solution advance. Unfortunately, I can't review this PR soon but I pinged some people to take a look on it.

Thanks and have a nice day too. : )

**Comment by ggbecker** [comment]
All right, I took a look at this PR and I see that it can be completely redesigned.

We only need a new file `Dockerfiles/test_suite-sle15` with the following content:

```
# This Dockerfile is a minimal example for a OL8-based SSG test suite target container.
FROM registry.suse.com/suse/sle15:15.3

ENV AUTH_KEYS=/root/.ssh/authorized_keys

ARG CLIENT_PUBLIC_KEY
ARG ADDITIONAL_PACKAGES

# Install Python so Ansible remediations can work
# Don't clean all, as the test scenario may require package install.
RUN zypper in --no-confirm openssh-server \
        openssh-clients \
        openscap \
        openscap-utils \
        tar \
        python3

RUN ssh-keygen -A -N '' -t ecdsa
RUN ssh-keygen -A -N '' -t rsa
RUN mkdir -p /root/.ssh
RUN printf "%s\n" "$CLIENT_PUBLIC_KEY" >> "$AUTH_KEYS"
RUN chmod og-rw /root/.ssh "$AUTH_KEYS"
```

And then you can build the `podman image` with the following command for example:

```
podman build --build-arg "CLIENT_PUBLIC_KEY=$(cat ~/.ssh/id_rsa.pub)" -t sle15 -f Dockerfiles/test_suite-sle15 .
```

Considering you have a public key located in `~/.ssh/id_rsa.pub`, or you can change to point to any other public key you have in your host. (This command works on `bash`)

This is all needed to build a podman container that can be used for automatus testing, then on your host you can build `sle15` content and run `automatus` with the following commands:

```
./build_product sle15 -d
python3 tests/automatus.py rule --remove-machine-only --container sle15 accounts_tmout --remediate-using ansible
```

It should run properly the tests.

Furthermore this file could be used in our CI similar with what we test here: https://github.com/ComplianceAsCode/content/blob/3a7694b8dc7e1e41435184f1448438a5751ee252/.github/workflows/automatus.yaml#L136

where --name would be `sle15` in this case. But it definitely opens the way to add sle15 to automated CI that runs automatus.

So my suggestion here is to re-do this pull request and include the minimum necessary to build a container that can run automatus tests. There is no need to build products as this should be done in the host, not from within the container.

@rumch-se I hope that this helps.

**Comment by ggbecker** [comment]
@rumch-se whenever you have time can you look at my comment above.

**Comment by rumch-se** [comment]
Hello @ggbecker 
I did the following steps
1. git clone from the current master on my host ~/ComplianceAsCode/content. Note I use Ubuntu 20.04 as a host OS.
2. creation of the file test_suite-sle15 into ~/ComplianceAsCode/content/Dockerfiles according to your recommendation. Note I replaced sle15:15.3 with sle15:15.4
3. execution of  the command podman build ... into ~/ComplianceAsCode/content according to your recommendation
4. execution of the command ./build_product sle15 -d and python3 tests/automatus.py  ... according to your recommendation

The attached file shows the output (error) which I had.
[execution_log.txt](https://github.com/ComplianceAsCode/content/files/11198974/execution_log.txt)

Have a nice day
Rumen

**Comment by ggbecker** [comment]
Running:

`zypper install -y libexpat1 libexpat-devel libexpat1 expat`

Should resolve the issue for now. As similar to:

https://github.com/ComplianceAsCode/content/blob/77c03101d5ef87466be27ced12c6bb816ca3bfe0/.github/workflows/gate.yaml#L43

**Comment by ggbecker** [comment]
@rumch-se Did you have the chance to test with the latest version. The `expat` dependency has been removed (https://github.com/ComplianceAsCode/content/pull/10467) from the project and it now should be simpler to run the commands.

My suggestion in this pull request is that can remove every file from it and keep only the Dockerfile I mentioned in few comments above, additionally you could create a new gating check that would run tests on this new container distro, for example: https://github.com/ComplianceAsCode/content/blob/master/.github/workflows/automatus.yaml

**Comment by ggbecker** [comment]
> 

@rumch-se this one should be almost good to go, if you can take a look and propose exactly the Dockerfile attached in one of the comments above.

**Comment by ggbecker** [comment]
@rumch-se ping

**Comment by rumch-se** [comment]
Hello @ggbecker 

Thank you for asking. 

I am working on this topic with a member of our team, and soon I will make a new commit. We have to make some additional tests...

Have a nice day
Rumen

**Comment by rumch-se** [comment]
Hello @ggbecker 

I have provided a new commit. Please review it.

Have a nice day
Rumen

File: Dockerfiles/sshd_config
Comment (ggbecker) [inline Dockerfiles/sshd_config:1]: I don't think this file is necessary for this PR to work. You can remove it.

File: Dockerfiles/test_suite-sle15
Comment (ggbecker) [inline Dockerfiles/test_suite-sle15:1]:
```suggestion
FROM registry.suse.com/suse/sle15:latest
```

Why not use always latest or simply not using anything after the colon?

**Comment by ggbecker** [comment]
The next step would be to use this file to run automatus tests on pull requests, same as we do in: https://github.com/ComplianceAsCode/content/blob/master/.github/workflows/automatus.yaml

**Comment by rumch-se** [comment]
Hello @ggbecker 
I removed the file 

I have the following general questions. 
Our team at SUSE wants to make some changes in github CI, which will provide:
- support of SLE 15 and SLE 12 instead of openSUSE
- improvement of unit tests 
- tests of ansible and bash remediation covering these two products. 

We are doing some research at the moment and I would like to know how these changes have to be delivered to this project. Who will review / approve them ? What is the usual practice in this case ?   

Have a nice day
Rumen

**Comment by ggbecker** [comment]
> Hello @ggbecker I removed the file
> 
> I have the following general questions. Our team at SUSE wants to make some changes in github CI, which will provide:
> 
>     * support of SLE 15 and SLE 12 instead of openSUSE
> 
>     * improvement of unit tests
> 
>     * tests of ansible and bash remediation covering these two products.
> 
> 
> We are doing some research at the moment and I would like to know how these changes have to be delivered to this project. Who will review / approve them ? What is the usual practice in this case ?
> 
> Have a nice day Rumen

The file I mentioned: https://github.com/ComplianceAsCode/content/blob/master/.github/workflows/automatus.yaml pretty much contains how to test ansible and bash remediation using container as a backend, that would be built by using this new Dockerfile you are providing. You just need to copy the file and adjust to SLE. The content-test-filtering project tries to detect which rule/product is being affected by a change in the pull request. In this case you probably want to hard code the product to be SLE and let the content-test-filtering to tell you which rules are being changed.

One limitation is that these tests run on a container backend, so rules that are applicable to bare metal will be skipped. On the other hand, we plan to add external resources (aka, virtual machines) to the github actions that will enable us run tests on those machines. It opens the ports for you to add machines to this pool as well.

If github actions support SLE containers, it should be easy to just switch from https://github.com/ComplianceAsCode/content/blob/c970c2257edad034f043675e923fd578141a8b4e/.github/workflows/gate.yaml#L40 to SLE for example.

**Comment by rumch-se** [comment]
Hello @ggbecker 

I have the following question. I understood that a new file for example automatus-sle15.yml has to be created and this file will take care about sle15. 

How can I test that the file is working ? I will need more details about this part.

Have a nice day
Rumen

**Comment by ggbecker** [comment]
> Hello @ggbecker
> 
> I have the following question. I understood that a new file for example automatus-sle15.yml has to be created and this file will take care about sle15.
> 
> How can I test that the file is working ? I will need more details about this part.
> 
> Have a nice day Rumen

The moment you propose a new file like that, it will start appearing as a new check in the github actions, and with that you can create a test commit on top of that to simulate a change in a rule. For example, change this to `no`: https://github.com/ComplianceAsCode/content/blob/2f1286b984fe40c6a526f82681389bccb956591d/linux_os/guide/services/ssh/ssh_server/sshd_enable_strictmodes/rule.yml#L61

With that the tests should be executed in your new workflow for SLE.

**Comment by rumch-se** [comment]
Hello @ggbecker 
With my last commit https://github.com/ComplianceAsCode/content/pull/9725/commits/fb85445fe618c5d20b232622ad54632f540fbbda I have provided support of Suse Linux Enterprise server SLE15. 
Some comments:
1. I modified the dockerfile - now it will support the official stable release SLE 15 SP5
2. I added a file automatus-SLE15 for which I used as reference the existing files for CS8 and CS9
3. I updated the file gate.yam by adding support for SLE 15. **Note:** At the moment the container in use does not install ShellCheck and I will need to discuss this topic with our maintainers from SUSE 

Have a nice day
Rumen

File: .github/workflows/automatus-sle15.yaml
Comment (ggbecker) [inline .github/workflows/automatus-sle15.yaml:25]:
```suggestion
        run: pip install json2html sphinxcontrib.jinjadomain GitPython deepdiff
```

File: .github/workflows/automatus-sle15.yaml
Comment (ggbecker) [inline .github/workflows/automatus-sle15.yaml:65]:
```suggestion
        run: ./build_product sle15
```

File: .github/workflows/automatus-sle15.yaml
Comment (ggbecker) [inline .github/workflows/automatus-sle15.yaml:12]: In theory you would not need to use a SLE distro to build the content. It could be the same from other workflows (I guess it's fedora), since the built artifacts should be same disregarding the distro. But it can be. I've added a few suggestions and will add a test commit to trigger a rule to be tested.

**Comment by rumch-se** [comment]
Hello @ggbecker 

I have modified gate.yml for SLE 15. Now it supports installation of bats and ShellCheck packages on SLE 15

I have a question, may I think that this PR provided what was expected to be delivered and if "yes" what are the plans for its integration into the master?

Have a nice day
Rumen

**Comment by rumch-se** [comment]
Hello @ggbecker 

Would be possible to progress on this PR? 

Have a nice day
Rumen

**Review by jan-cerny — APPROVED** [review]
I think that as we optimized the GitHub CI recently, we can now afford to add a new job. However, we should continue optimizing further in future PRs.

**Pull request merged**
